### PR TITLE
Add SNMP KDF

### PR DESCRIPTION
**Pull request by bfussell**
This PR adds support for SNMP KDF compliant with SP800-135, RFC 3414 and RFC 7860.
The SNMP KDF will be available in the default and FIPS providers.
CLA submitted via email.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (paulidale) [inline doc/man7/EVP_KDF-SNMPKDF.pod:38]: Try this instead: _This parameter sets the snmpEngineID value for the KDF._

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (paulidale) [inline doc/man7/EVP_KDF-SNMPKDF.pod:52]: allocate rather than alloc.

File: include/openssl/kdf.h
Comment (paulidale) [inline include/openssl/kdf.h:66]: Is this relevant for a public header?

File: providers/fips/fipsprov.c
Comment (paulidale) [inline providers/fips/fipsprov.c:387]: `FIPS_DEFAULT_PROPERTIES` rather than `"provider=default"`.
This isn't the default provider after all.

I'm not sure what policy we should adopt for items that can be FIPS approved but aren't in any of our current validations.

File: providers/implementations/kdfs/snmpkdf.c
Comment (paulidale) [inline providers/implementations/kdfs/snmpkdf.c:2]: 2023 I presume

File: providers/implementations/kdfs/snmpkdf.c
Comment (paulidale) [inline providers/implementations/kdfs/snmpkdf.c:196]: Our coding style is different for block comments.
```c
/*
 * SNMPKDF - In compliance with SP800-135 and RFC 7860, calculate
 * …
 */
```

File: test/evp_kdf_test.c
Comment (paulidale) [inline test/evp_kdf_test.c:1443]: Are the tests added here exercising anything that cannot be done via evp_test?

If they are, they are fine.  If not, they'd be better as part of evp_test.

Superficially, it looks like they might be.

**Review by paulidale — CHANGES_REQUESTED** [review]
Overall, this looks quite good.

**Review by paulidale — COMMENTED** [review]
We'd normally also include a CTX_dup call which appears to be missing.

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (bfussell) [reply doc/man7/EVP_KDF-SNMPKDF.pod]: done

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (bfussell) [reply doc/man7/EVP_KDF-SNMPKDF.pod]: done

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (bfussell) [reply doc/man7/EVP_KDF-SNMPKDF.pod]: changed to pass, replaced with proper length, no size_t

File: include/openssl/kdf.h
Comment (bfussell) [reply include/openssl/kdf.h]: Nope, moved to snmpkdf.c

File: providers/fips/fipsprov.c
Comment (bfussell) [reply providers/fips/fipsprov.c]: Changed to FIPS_DEFAULT_PROPERTIES

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: Addressed

**Comment by bfussell** [comment]
Added the CTX_dup function.

**Comment by bfussell** [comment]
Re-Open after CLA approved.

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (slontis) [inline doc/man7/EVP_KDF-SNMPKDF.pod:76]: use sizeof().. These magic numbers dont match the arrays above..,

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (slontis) [inline doc/man7/EVP_KDF-SNMPKDF.pod:100]: Not sure what we should be putting here since this wont make it into 3.2.
 @openssl/otc is this going to be 3.3?

File: include/openssl/kdf.h
Comment (slontis) [inline include/openssl/kdf.h:66]: revert this file..

File: providers/fips/self_test_data.inc
Comment (slontis) [inline providers/fips/self_test_data.inc:389]: Maybe reference where this comes from in a comment..

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:1637]: unless it is necessary I would have done the failure tests as evpkdf_snmp.txt entries. You can use the Result tag e..g.
````
Result = KDF_DERIVE_ERROR
````

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:42]: indent 1 more...

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:237]: follows

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:245]: okey?

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:260]: How does this limit to SHA1/SHA2? (e.g. not use  SHA3)..
Also is the legacy case for MD5 required in non fips mode?

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:265]: != NULL

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:262]: Remove {} around single line statements.

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:271]: <= ?

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:274]: Use reset as init this is just an alias. I dont think the call is needed straight after a new

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:278]: Shouldnt password_len be limited in set_params to < KDF_SNMP_PASSWORD_HASH_AMOUNT

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:306]: Using okeylen looks dangerous if it is not checked against the digest size.

**Comment by slontis** [comment]
800-135r1 only lists SHA1.. Are there other places that specify that other digests can be used in fips mode (If not it probably needs a separate check inside a ````#ifdef FIPS_MODULE```` block,,

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (bfussell) [reply doc/man7/EVP_KDF-SNMPKDF.pod]: done

File: include/openssl/kdf.h
Comment (bfussell) [reply include/openssl/kdf.h]: done

File: providers/fips/self_test_data.inc
Comment (bfussell) [reply providers/fips/self_test_data.inc:392]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: test/evp_kdf_test.c
Comment (bfussell) [reply test/evp_kdf_test.c:1637]: moved tests that could be done in evpkdf_snmp.txt

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: == NULL

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: removed

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c:302]: added check

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c:275]: Minimum is 64 according to NIST as well as RFC3414

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: removed check to allow other hash types

**Comment by bfussell** [comment]
> 800-135r1 only lists SHA1.. Are there other places that specify that other digests can be used in fips mode (If not it probably needs a separate check inside a `#ifdef FIPS_MODULE` block,,

SHA1 is the only hash currently supported in FIPS mode, however there is interest in using SHA2. We have some products that choose to use SHA2 for the KDF and list it as non-approved. I can add the #ifdef if you'd like to be strict. Are all the KDFs supposed to restrict hashes to FIPS only values for the FIPS provider or are they controlled by policy ?

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (slontis) [inline doc/man7/EVP_KDF-SNMPKDF.pod:74]: Update the comment on line 57 to match this..

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:1671]: I am wondering if the check for a NULL key should be moved up to the EVP_KDF_derive().. Does it make sense for any KDF to allow this to be NULL?
I can see existing KDF's that are not checking this at the lower level so they would crash currently.

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (slontis) [inline doc/man7/EVP_KDF-SNMPKDF.pod:50]: This statement seems confusing (or You are missing some words).
If you are referring to the OSSL_PARAM [] this can be NULL. since there is also a seperate EVP_KDF_CTX_set_params() that can be called.

Are you saying that the maximum length of the output I<keylen> is the size of the specified digest?

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:142]: indent off by 1

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:163]: move || down

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [reply providers/implementations/kdfs/snmpkdf.c:275]: Where is the minimum bound check?

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:1640]: Use TEST_ptr()

File: test/evp_kdf_test.c
Comment (slontis) [inline test/evp_kdf_test.c:1690]: Is this going to do an out of bounds read?

File: test/evp_kdf_test.c
Comment (bfussell) [reply test/evp_kdf_test.c:1671]: I don't see a reason for that. Moving the check into EVP_KDF_derive would certainly catch it earlier and cover all KDFs.

File: test/evp_kdf_test.c
Comment (bfussell) [reply test/evp_kdf_test.c:1695]: fixed

File: test/evp_kdf_test.c
Comment (bfussell) [reply test/evp_kdf_test.c]: fixed

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: fixed

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: fixed

File: doc/man7/EVP_KDF-SNMPKDF.pod
Comment (bfussell) [reply doc/man7/EVP_KDF-SNMPKDF.pod]: Rewritten

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c:278]: added

**Comment by bfussell** [comment]
@slontis @paulidale Any further updates/changes you're expecting from me on this PR ?

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:116]: I know this was copied from somewhere else, but can you see what happens for the cases of p->data_size = 0
OR p->data = NULL.

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:217]: use OSSL_DISPATCH_END

File: test/recipes/30-test_evp_data/evpkdf_snmp.txt
Comment (slontis) [inline test/recipes/30-test_evp_data/evpkdf_snmp.txt:16]: You may need to add the following line before each KDF = SNMPKDF
FIPSversion = >=3.3.0


There are backwards compatibility tests for old versions of OpenSSL FIPS providers.
This line indicates only to test against new FIPS providers.

**Review by slontis — CHANGES_REQUESTED** [review]

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [reply providers/implementations/kdfs/snmpkdf.c:116]: I assume this was allowing some settings to optionally be set back to NULL. Probably not needed if the values need to be set.

Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c:116]: p->data = NULL triggers snmpkdf_set_membuf to return 0 indicating an error with the params
p->data_size = 0 is allowed

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

File: test/recipes/30-test_evp_data/evpkdf_snmp.txt
Comment (bfussell) [reply test/recipes/30-test_evp_data/evpkdf_snmp.txt:17]: done

File: providers/implementations/kdfs/snmpkdf.c
Comment (slontis) [inline providers/implementations/kdfs/snmpkdf.c:288]: Normally this code would be collapsed..

````
if (!EVP_DigestInit_ex(md, evp_md, NULL)
        || !
        || !  )
    goto err;
 ````

**Comment by slontis** [comment]
ping @openssl/committers

**Comment by t8m** [comment]
@bfussell Could you please rebase this to up-to-date master branch, drop the merge commits and squash the commits to one with an appropriate commit message?

File: providers/implementations/kdfs/snmpkdf.c
Comment (bfussell) [reply providers/implementations/kdfs/snmpkdf.c]: done

**Comment by t8m** [comment]
Please use `git rebase -i  master` to rebase against fresh master branch. With that command you can squash the commits into a single one and drop the merge commits.

**Comment by bfussell** [comment]
@t8m I've tried everything I know to get the commits to squash with no luck.  On our local git repo it works fine. Any suggestions on alternatives ? Create a new PR with the full commit possibly ?

**Comment by t8m** [comment]
> @t8m I've tried everything I know to get the commits to squash with no luck. On our local git repo it works fine. Any suggestions on alternatives ? Create a new PR with the full commit possibly ?

I think one problem that might have caused this is that you did not create a new branch but instead used the master branch in the bfussell/openssl repository. So please create a new PR but this time create a new branch for this work.

**Comment by xnox** [comment]
This has been implemented in https://github.com/openssl/openssl/commit/1b035166bdb2d86e718ad51a55caffc8bc9504a7

**Pull request closed without merging**
